core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/s3/S3FileMetadataSource.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+
+package co.cask.hydrator.plugin.batch.file.s3;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Macro;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.batch.Input;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.batch.BatchRuntimeContext;
+import co.cask.cdap.etl.api.batch.BatchSource;
+import co.cask.cdap.etl.api.batch.BatchSourceContext;
+import co.cask.hydrator.common.SourceInputFormatProvider;
+import co.cask.hydrator.common.batch.JobUtils;
+import co.cask.hydrator.plugin.batch.file.AbstractFileMetadata;
+import co.cask.hydrator.plugin.batch.file.AbstractFileMetadataSource;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.Job;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * FileCopySource plugin that pulls filemetadata from S3 Filesystem.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name("S3FileMetadataSource")
+@Description("Reads file metadata from S3 bucket.")
+public class S3FileMetadataSource extends AbstractFileMetadataSource<S3FileMetadata> {
+  private S3FileMetadataSourceConfig config;
+  private static final Logger LOG = LoggerFactory.getLogger(S3FileMetadataSource.class);
+
+  public S3FileMetadataSource(S3FileMetadataSourceConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  public static final Schema S3_CREDENTIAL_SCHEMA = Schema.recordOf(
+    "metadata",
+    Schema.Field.of(S3FileMetadata.S3Credentials.DATA_BASE_TYPE, Schema.of(Schema.Type.STRING)),
+    Schema.Field.of(S3FileMetadata.S3Credentials.ACCESS_KEY_ID, Schema.of(Schema.Type.STRING)),
+    Schema.Field.of(S3FileMetadata.S3Credentials.SECRET_KEY_ID, Schema.of(Schema.Type.STRING)),
+    Schema.Field.of(S3FileMetadata.S3Credentials.REGION, Schema.of(Schema.Type.STRING)),
+    Schema.Field.of(S3FileMetadata.S3Credentials.BUCKET_NAME, Schema.of(Schema.Type.STRING))
+  );
+
+  private Schema outputSchema;
+
```

Comment:
there should be some validation done on the config in the configurePipeline method. This ensures that invalid configs cause the pipeline to fail to deploy instead of waiting until runtime to fail.